### ISSUE TITLE
утечка памяти в tvb_pf_payment (positive cash)

### ISSUE THREAD

**Issue by alexandra-gsoftware** [gsbelarus/check-and-cash#243]
при оплате чека создается объект класса tvb_pf_payment . В нем дизайнером создается форма оплаты. Объект уничтожается только при закрытии формы кассира, соответственно, форма оплаты создается столько раз, сколько было закрыто чеков, а потом все эти одинаковые формы одновременно уничтожаются (если программа не крашнется до этого).

**Comment by hirurg-lybitel** [comment]
Так в деструкторе форма удаляется же.

**Comment by alexandra-gsoftware** [comment]
но деструктор вызывается, когда закрывается окно кассира, а не когда заканчивается скрипт

**Comment by hirurg-lybitel** [comment]
почему?
По идее же объект класса удаляется после окончани функции его создавшей.

**Comment by alexandra-gsoftware** [comment]
Хрен знает. Возможно, он не может уничтожиться, тк он связан с формой кассира через Transaction. Пока транзакшн жива, он не может уничтожиться. Если поставить точки остановки в конструкторе и в деструкторе, то деструктор вызывается столько раз, сколько было пробито чеков, в момент закрытия формы кассира

**Comment by hirurg-lybitel** [comment]
Создание форм никак не связано с транзакциями.

Видимо, особенность местного компилятора.
Добавлю тогда вручную уничтожение объекта класса.

**Comment by alexandra-gsoftware** [comment]
Форма не при чем. Я думаю, из-за вот этой строки
Set m_Transaction = Transaction

Там в середине еще несколько exit function есть, тогда уж перед ними тоже ее надо убивать объект

**Comment by hirurg-lybitel** [comment]
Не вижу никакой проблемы в Set m_Transaction = Transaction. 
Это просто присваивание внутренней переменной, которая никоим образом не учитывается при удалении экземпляра класса.

Вероятно всё, что создаётся не через TCreator ведёт себя не совсем понятно. Точнее как-то иначе.

**Comment by alexandra-gsoftware** [comment]
не, ну TCreator  ведь именно так работает, как ты предполагаешь. И другие классы. Если бы объект уничтожался при закрытии Гедымина, можно было бы списать на "конструктор с параметрами работает не так, как конструктор без параметров". А он уничтожается при закрыти какой-то левой формы.
Кстати, если закомментировать передачу и использование транзакшн, деструктор все равно вызывается только при закрытии формы кассира

**Comment by alexandra-gsoftware** [comment]
при ручном уничтожении объекта - работает

**Comment by gsbelarus** [comment]
Все, что создается через дизайнер надо **обязательно** руками удалять. Хотите автоматического удаления -- используйте TCreator. Никакой магии в нем нет, он регистрирует в списке все, созданные с его помощью объекты и при своем удалении удалит через дизайнер эти объекты. Не используете TCreator -- вся ответственность удаления объектов лежит на вас.

**Comment by gsbelarus** [comment]
Если кто еще вспомнит по коду такие места -- их надо поправить.

Вообще, надо понимать, что в скриптах мы не можем работать с Делфи объектами. Поэтому мы создаем COM-объекты обертки вокруг делфи объектов. Обертка не делает ничего, кроме как хранит ссылку на делфовский объект и предоставляет в скрипт набор методов/свойств. Когда они вызываются -- вызываются соответствующие методы/свойста делфи объекта.

Возникает вопрос, почему бы при уничтожении обертки не удалять делфи объект?

Во-первых, этот объект может принадлежать чему-то другому. Например, свойство типа TStringList в каком-нибудь компоненте. Если мы ему вызовем принудительно .Free, то дальше компонент просто упадет с AV, когда попытается обратиться к этому списку.

Во-вторых, мы сами можем передать этот объект в макросе в другой объект, где он будет использоваться внутри Делфи кода. Будет совсем нехорошо, если в процессе использования объект вдруг удалится, так как соответствующий ему объект-оболочка больше не нужен VBScript.

Поэтому у нас не было другого кроме как предоставить разработчику два варианта: дизайнер, чтобы создавать/удалять объекты вручную, под свою ответственность, и TCreator в тех случаях, когда известно, что это наш объект, мы его создали и дальше вот-этой области он не уйдет.

Собственно, если бы мы в свое время использовали не VBScript, а JScript, то потребности в TCreator не было бы. Разработчик бы использовал конструкции `try... catch ... finally` в сочетании с дизайнером. Но, в VBScript таких конструкций нет, и соответстенно TCreator единственный относительно простой способ высвобождения ресурсов в случае возникновения исключения.

**Comment by alexandra-gsoftware** [comment]
@gsbelarus 
поясняю ситуацию. Есть VB класс. Объект этого класса создается в некой функции. Но при завершении этой функции объект не уничтожается. Он уничтожается, когда уничтожается форма, которой принадлежит событие, которое вызывает эту функцию. Почему так - непонятно.

**Comment by gsbelarus** [comment]
дайте этот код сюда.

**Comment by alexandra-gsoftware** [comment]
короче. если конструктор сделать обычной функцией и переоткрыть гедымин, то объект уничтожится сам

**Comment by gsbelarus** [comment]
тут все просто:

1. область использования объекта -- это функция или процедура
2. каждый объект имеет связанный с ним счетчик ссылок
3. объекты в область или попадают извне (как параметры функции или процедуры), или создаются внутри оператором `New`
4. счетчик созданного объекта устанавливается в 1
5. когда объект передается вне области (например, как параметр в другую функцию или присваивается как результат функции) его счетчик инкрементируется
6. когда область заканчивается (выходим из функции или процедуры) счетчик ссылок объекта декрементируется
7. оператор `Set Obj = Nothing` тоже декрементирует счетчик данного объекта
8. объекты, счетчики которых стали равны нулю, удаляются

В данном примере объект создавался оператором `New` -- счетчик = 1, а затем возвращался с использованием ключевого слова `me` из метода `Init` этого же объекта как результат, что приводило к еще одному увеличению счетчика ссылок:
```
  Function Init
    ...
    Set Init = me
  End Function
```
Таким образом, при достижении конца функции счетчик уменьшался, но оставался большим нуля.

**Comment by alexandra-gsoftware** [comment]
Хм, так это баг отладчика?

**Comment by hirurg-lybitel** [comment]
> тут все просто:
> 
> 1. область использования объекта -- это функция или процедура
> 2. каждый объект имеет связанный с ним счетчик ссылок
> 3. объекты в область или попадают извне (как параметры функции или процедуры), или создаются внутри оператором `New`
> 4. счетчик созданного объекта устанавливается в 1
> 5. когда объект передается вне области (например, как параметр в другую функцию или присваивается как результат функции) его счетчик инкрементируется
> 6. когда область заканчивается (выходим из функции или процедуры) счетчик ссылок объекта декрементируется
> 7. оператор `Set Obj = Nothing` тоже декрементирует счетчик данного объекта
> 8. объекты, счетчики которых стали равны нулю, удаляются
> 
> В данном примере объект создавался оператором `New` -- счетчик = 1, а затем возвращался с использованием ключевого слова `me` из метода `Init` этого же объекта как результат, что приводило к еще одному увеличению счетчика ссылок:
> 
> ```
>   Function Init
>     ...
>     Set Init = me
>   End Function
> ```
> 
> Таким образом, при достижении конца функции счетчик уменьшался, но оставался большим нуля.
> 
> PS: Похоже, что в такой ситуации:
> 
> ```
>   Class C
>     Public Default Function Init(A, B, C)
>       ...
>       Set Init = me
>     End Function
>   End Class
> 
>   Set Obj = (New C)(1, 1, 1)
> ```
> 
> Счетчик Obj устанавливается в 2.

Так конструкция 
```
 Class C
    Public Default Function Init(A, B, C)
      ...
      Set Init = me
    End Function
  End Class
```

неприменима или что? У нас же память так утекать всё время будет.

**Comment by gsbelarus** [comment]
да. неприменимо. me нужен если тебе изнутри класса надо вызвать другую функцию, куда следует передать экземпляр ТЕКУЩЕГО объекта. На входе в такую функцию внутренний счетчик инкрементируется, а по ее завершении -- декрементируется.

Тут же идет инкремент и объект уходит наружу и никто не знает когда в нем отпадет необходимость.

Это логическая ошибка ведущая у течке памяти. Надо переделать код.

**Comment by gsbelarus** [comment]
Почему просто не создавать объект оператором New а потом вызывать у него какую нибудь функцию инициализации внутренних параметров.

**Comment by hirurg-lybitel** [comment]
> да. неприменимо. me нужен если тебе изнутри класса надо вызвать другую функцию, куда следует передать экземпляр ТЕКУЩЕГО объекта. На входе в такую функцию внутренний счетчик инкрементируется, а по ее завершении -- декрементируется.
> 
> Тут же идет инкремент и объект уходит наружу и никто не знает когда в нем отпадет необходимость.
> 
> Это логическая ошибка ведущая у течке памяти. Надо переделать код.

Код уже давно переделан. Проблема не в нём. Это просто уточняющий вопрос.




> Почему просто не создавать объект оператором New а потом вызывать у него какую нибудь функцию инициализации внутренних параметров.

Попытка создать интерфейс для обычного создания класса с параметризированным конструктором, как привыкли.
